packages/physics-physx/src/PhysXPhysicsManager.ts

```diff
@@ -0,0 +1,253 @@
+ */
+export class PhysXPhysicsManager implements IPhysicsManager {
+  private static _tempPosition: Vector3 = new Vector3();
+  private static _tempNormal: Vector3 = new Vector3();
+  private static _pxRaycastHit: any;
+  private static _pxFilterData: any;
+  private static _tempGravity = {
+    x: 0,
+    y: -9.81,
+    z: 0
+  };
+  private static _tempOrigin = {
+    x: 0,
+    y: 0,
+    z: 0
+  };
+  private static _tempDirection = {
+    x: 0,
+    y: 0,
+    z: 0
+  };
+
+  private _queryFlag: QueryFlag = QueryFlag.STATIC | QueryFlag.DYNAMIC;
+  private _pxScene: any;
+
+  private readonly _onContactBegin?: Function;
+  private readonly _onContactEnd?: Function;
+  private readonly _onContactPersist?: Function;
+  private readonly _onTriggerBegin?: Function;
+  private readonly _onTriggerEnd?: Function;
+  private readonly _onTriggerPersist?: Function;
+
+  private _eventMap: Map<number, [number, PhysicsState]> = new Map();
+
+  private _gravity: Vector3 = new Vector3(0, -9.81, 0);
+
+  /**
+   * Global gravity in the physical scene.
+   */
+  get gravity(): Vector3 {
+    return this._gravity;
+  }
+
+  set gravity(value: Vector3) {
+    if (this._gravity !== value) {
+      value.cloneTo(this._gravity);
+    }
+    PhysXPhysicsManager._tempGravity.x = value.x;
+    PhysXPhysicsManager._tempGravity.y = value.y;
+    PhysXPhysicsManager._tempGravity.z = value.z;
+    this._pxScene.setGravity(PhysXPhysicsManager._tempGravity);
+  }
+
+  constructor(
+    onContactBegin?: (obj1: number, obj2: number) => void,
+    onContactEnd?: (obj1: number, obj2: number) => void,
+    onContactPersist?: (obj1: number, obj2: number) => void,
+    onTriggerBegin?: (obj1: number, obj2: number) => void,
+    onTriggerEnd?: (obj1: number, obj2: number) => void,
+    onTriggerPersist?: (obj1: number, obj2: number) => void
+  ) {
+    this._onContactBegin = onContactBegin;
+    this._onContactEnd = onContactEnd;
+    this._onContactPersist = onContactPersist;
+    this._onTriggerBegin = onTriggerBegin;
+    this._onTriggerEnd = onTriggerEnd;
+    this._onTriggerPersist = onTriggerPersist;
+
+    const triggerCallback = {
+      onContactBegin: (obj1, obj2) => {},
+      onContactEnd: (obj1, obj2) => {},
+      onContactPersist: (obj1, obj2) => {},
+      onTriggerBegin: (obj1, obj2) => {
+        this._eventMap.set(obj1.getQueryFilterData().word0, [
+          obj2.getQueryFilterData().word0,
+          PhysicsState.TOUCH_FOUND
+        ]);
+        this._eventMap.set(obj2.getQueryFilterData().word0, [
+          obj1.getQueryFilterData().word0,
+          PhysicsState.TOUCH_FOUND
+        ]);
+      },
+      onTriggerEnd: (obj1, obj2) => {
+        this._eventMap.set(obj1.getQueryFilterData().word0, [obj2.getQueryFilterData().word0, PhysicsState.TOUCH_LOST]);
+        this._eventMap.set(obj2.getQueryFilterData().word0, [obj1.getQueryFilterData().word0, PhysicsState.TOUCH_LOST]);
+      }
+    };
```

Comment:
You can use `Three-tier data structure`:
```ts 
 private _eventCollisons: Collision[] = []; // Only store current frame Collisions which fire event.
 private _collisonMap: Record<number, Record<number, Collision>> = {};// Use object to store collision for easy to find
 private _collisonPool: CollisionPool;// Use collisionPool to ignore GC and save memory.
```

Reply:
done